graph/src/blockchain/firehose_block_ingestor.rs

```diff
@@ -82,6 +82,87 @@ where
         }
     }
 
+    pub async fn run_backfill(self) {
+        use firehose::ForkStep::*;
+
+        let mut backoff =
+            ExponentialBackoff::new(Duration::from_millis(250), Duration::from_secs(30));
+
+        let mut backfill_cursor = self.fetch_backfill_cursor().await;
+
+        loop {
+            let backfill_completed = self.fetch_backfill_is_completed().await;
+            if backfill_completed {
+                info!(self.logger, "Backfill completed");
+                break;
+            }
+
+            let backfill_target_block_number = self.fetch_backfill_target_block_num().await;
+            let result = self
+                .endpoint
+                .clone()
+                .stream_blocks(firehose::Request {
+                    start_block_num: 0,
+                    stop_block_num: backfill_target_block_number as u64,
+                    start_cursor: backfill_cursor.clone(),
+                    fork_steps: vec![StepIrreversible as i32],
+                    ..Default::default()
+                })
+                .await;
+
+            match result {
+                Ok(stream) => {
+                    backfill_cursor = self
+                        .process_backfill_blocks(
+                            backfill_cursor,
+                            stream,
+                            backfill_target_block_number,
+                        )
+                        .await
+                }
+                Err(e) => {
+                    error!(
+                        self.logger,
+                        "Unable to connect to backfill endpoint: {:?}", e
+                    )
+                }
+            }
+
+            // If we reach this point, we must wait a bit before retrying
+            backoff.sleep_async().await;
+        }
+    }
+
+    async fn fetch_backfill_is_completed(&self) -> bool {
+        let mut backoff =
+            ExponentialBackoff::new(Duration::from_millis(250), Duration::from_secs(30));
+        loop {
+            match self.chain_store.clone().chain_backfill_is_completed() {
```

Comment:
I feel `chain_backfill_is_completed` should return `Result<bool, Error>`. What the case where it can be `Ok(None)` exactly? Wouldn't it better to return in `Error` if backfill completeness can't be determined somehow?

Reply:
done!